### DISCUSSION TITLE
초롱의 우당탕탕 인프라 기록일지

### DISCUSSION THREAD

**Discussion by HoeSeong123** [woowacourse-teams/2024-code-zap#426] [기록해봐요]
# vscode로 EC2에 쉽게 접근하는 방법!!
## ❗주의: 해당 방법은 서버에 부하가 걸려서 권장되지 않는 방법입니다. `t4g.micro`에서 사용하면 서버 멈출 수도 있어요..
1.Extensions 클릭 후 `SSH` 검색 및 설치!!
![image](https://github.com/user-attachments/assets/a40a51f5-3e0c-4698-aaed-5d28b65f69b7)

2. config 파일 열기
![image](https://github.com/user-attachments/assets/1f5f9abe-7315-4ff8-8a57-9b00a1cb52bf)

3. 아래의 코드 추가하기
```
Host code-zap-develop
    HostName 3.34.226.185
    User ubuntu
    IdentityFile /Users/hoeseong/key-code-zap.pem
    Port 22

Host code-zap-dev-db
    HostName 10.0.100.197
    User ubuntu
    IdentityFile /Users/hoeseong/key-code-zap.pem
    Port 22
    ProxyJump code-zap-develop
```
- `Host`: 내가 지정한 이름
- `HostName`: 접근하고자 하는 인스턴스의 ip 주소
- `User`: 계정 이름
- `Port`: 연결할 포트 번호, 기본포트(22번)일 경우 생략 가능
- `IdentityFile`: 키 파일(.pem 확장자) 위치
- `ProxyJump`: 우회해서 접속할 인스턴스의 이름 (위의 `Host`에서 지정한 이름)

4. 저장 및 새로고침 후 원하는 인스턴스에 접속
![image](https://github.com/user-attachments/assets/cf13a0aa-77ac-433c-b607-c642d78fe03b)
<img width="1012" alt="image" src="https://github.com/user-attachments/assets/5a92ca81-7147-4c84-80df-8bf605098a0a">

**Comment by HoeSeong123** [comment]
1. develop EC2 안에 있는 도커에서 sql 데이터 백업하기 ->`.sql` 파일로 저장
    `mysqldump -u root -p<비밀번호>  <백업할 데이터베이스> > <백업할 파일명>`
    -> `mysqldump -u root -ppassword code_zap > code_zap.sql`

3. 도커에서 로컬로 파일 꺼내기 
    `docker cp <컨테이너 이름>:<컨테이너 내부 파일 경로> <복사할 파일 경로> ` 
    -> `docker cp code-zap-db:/code_zap.sql ~/download`

4. EC2 로컬에서 내 노트북 로컬로 파일 옮기기 
    `scp <EC2 Host명>:<저장된 파일 경로> <복사할 파일 경로>` 
    -> `scp code-zap-develop:~/download/code_zap.sql ~/`

5. 내 노트북 로컬에서 dev db EC2로 파일 옮기기
    `scp <복사할 파일 경로> <EC2 Host명>:<저장된 파일 경로>` 
    -> `scp code_zap.sql code-zap-dev-db:~/code-zap`

6. EC2 로컬에서 도커로 파일 옮기기
    `docker exec -i <컨테이너 이름> mysql -u root -p<비밀번호> <적용할 데이터베이스> < <백업 파일명>` 
    -> `docker exec -i code-zap mysql -u root -ppassword code_zap < code_zap.sql`

**Reply by zangsu** [reply]
6. EC2 로컬에서 내 노트북 로컬로 "**폴더**" 옮기기
scp -r <EC2 Host명>:<저장된 폴더 경로> <복사할 폴더 경로>
-> `scp -r zap-dev:~/srv/code-zap/script ~/Desktop/codezap/srv`

**Comment by HoeSeong123** [comment]
# 한글 전부 깨짐
<img width="870" alt="image" src="https://github.com/user-attachments/assets/dcab60f1-ab40-4d73-9aac-b1d1d5b0138d">
기껏 db 백업해서 가져왔더니 한글이 전부 깨지는 현상이 발생함

## status
<img width="859" alt="스크린샷 2024-08-14 오후 6 12 42" src="https://github.com/user-attachments/assets/56b054a6-9fcf-4045-90a9-68b619003390">
검색해보니 `status` 명령어를 통해 확인해보라고 해서 확인해봣더니 `Client characterset`과 `Conn. characterset`이 utf8이 아닌 latin1로 되어 있는 것을 확인했다.

한글 설정을 위해 아래의 설정까지 해줬는데 어째서인지 되지 않았다.
<img width="321" alt="스크린샷 2024-08-14 오후 6 16 06" src="https://github.com/user-attachments/assets/09c95d81-efd3-4bf7-bd85-9b90b3d37fbe">

아래의 링크들에서 해결책을 찾을 수 있었다.
[링크 1](https://restato.github.io/posts/mysql-hangul-encoding/)
[링크 2](https://ink1234.tistory.com/14)

저 설정으로는 Server와 DB 부분만 설정이 가능하고 Client와 conn은 my.cnf 파일을 직접 수정함으로써 가능하다는 것이었다.

컨테이너 안의 파일들은 수정이 힘들지만 `volumes`라는 엄청난 설정을 통해 수정이 가능했다.
### volumes란
volumes에 대해 간단히 설명하면 도커 내부의 폴더를 로컬에 그대로 연결하는 기능? 이라고 설명할 수 있을 것 같다.
추후에 더 자세히 조사해보겠음

우리는 아래와 같은 설정을 해주었다.
```
volumes:
  - ./data:/var/lib/mysql
  - ./config:/etc/mysql/conf.d
```
간단히 설명해보자면 도커 내부의 `/var/lib/mysql` 폴더들을 우리 로컬의 `/data`에 연결한다는 뜻이다.
마찬가지로 `/etc/mysql/conf.d` 폴더들은 우리 로컬의 `/config`에 연결된다.

로컬에서 해당 폴더에 들어가 파일을 추가하는 등의 작업을 하면 도커를 실행할 때 그대로 복사가 된다.
따라서 나는 위 [링크 1](https://restato.github.io/posts/mysql-hangul-encoding/)을 따라서 `/config` 폴더에 들어가 `my.cnf` 파일을 생성한 후 아래 설정을 추가해주었다.

```
[client]
default-character-set=utf8
 
[mysql]
default-character-set=utf8
 
[mysqld]
collation-server = utf8_unicode_ci
character-set-server = utf8
```

그랬더니 아래처럼 utf8로 잘 변경이 된 것을 확인할 수 있었고
<img width="862" alt="스크린샷 2024-08-14 오후 6 24 09" src="https://github.com/user-attachments/assets/a107bfab-9743-4356-94d8-4ec5dbd2533b">

아래와 같이 한글도 잘 출력되는 것을 확인할 수 있었다!!
<img width="865" alt="image" src="https://github.com/user-attachments/assets/1197e532-4a1e-429e-9355-bdf7ed3cc503">

**Comment by HoeSeong123** [comment]
# Nginx 파헤치기
## 설정파일
첫 관문부터 난관이었다. etc/nginx에 설정 파일이 있다는데 눈을 씻고 찾아봐도 그런 폴더는 보이지 않았다.

<img width="515" alt="스크린샷 2024-08-15 오후 3 14 04" src="https://github.com/user-attachments/assets/fba824d7-a2e7-480e-bd22-34a0d85c0a14">


알고보니 `~/` 경로에 있는게 아니라 거기보다 더 아래인 `/` 경로에 있던 파일이었다...
<img width="488" alt="스크린샷 2024-08-15 오후 3 15 43" src="https://github.com/user-attachments/assets/7d74c407-3567-4406-a790-bdc712ca98fe">

이제 `nginx.conf` 파일에 들어가서 설정 파일을 뜯어보기 시작했다.
고 하자마자 또 난관에 봉착.
```
location / {
	root 404;
	proxy_pass $dev_url;
	proxy_set_header Host $http_host;
	proxy_set_header X-Real-IP $remote_addr;
	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
	proxy_set_header X-Forwarded-Proto $scheme;
}
```
저 `$dev_url`은 어디에 있는 변수지...

이 변수들은 `etc/nginx/conf.d/code-zap-dev.inc`에 있는 변수인데 아래의 설정을 통해 들어오게 된 값들이다.
```
include /etc/nginx/conf.d/code-zap-dev.inc;
```
그래서 해당 파일에 들어가면 아래와 같은 코드가 작성되어 있다.
```
set $dev_url http://127.0.0.1:8084;
```
~~(다들 알고 있겠지만 127.0.0.1은 localhost를 뜻한다. 나는 이번에 처음 알았다)~~
